### PR TITLE
add debug assertion to `unreachable_unchecked`

### PR DESCRIPTION
**Pull request by ibraheemdev**

### CODE REVIEW

**Comment by joshtriplett** [comment]
I'd like to understand better whether this specific `debug_assert` will get compiled out of programs when built in release mode.

Would it make sense for us to make this `#[inline(always)]` to ensure that?

**Comment by Mark-Simulacrum** [comment]
It wouldn't even get codegen'd - std is not shipped with debug assertions on, so it shouldn't get generated anywhere.

I'd probably implement this as an explicit cfg(...) around a panic, though, rather than debug_assert with a false argument.

**Comment by ibraheemdev** [comment]
```
error: `panic` is not yet stable as a const fn
    |
23  | / pub macro panic_2015 {
24  | |     () => (
25  | |         $crate::panicking::panic("explicit panic")
---
12  | |     };
13  | | }
    | |_- in this expansion of `panic!` (#2)
...
214 | / macro_rules! debug_assert {
215 | |     ($($arg:tt)*) => (if $crate::cfg!(debug_assertions) { $crate::assert!($($arg)*); })
    | |                                                           |
    | |                                                           in this macro invocation (#2)
    | |                                                           in this macro invocation (#3)
216 | | }
216 | | }
```

Why is this failing? `unreachable_unchecked` is const-stable from 1.57, as of which const-panic is stable, right?

**Comment by camsteffen** [comment]
How about `#[cfg_attr(debug_assertions, track_caller)]` as well?

File: library/core/src/hint.rs [hunk 1]
```diff
@@ -46,6 +46,7 @@ use crate::intrinsics;
 #[stable(feature = "unreachable", since = "1.27.0")]
 #[rustc_const_stable(feature = "const_unreachable_unchecked", since = "1.57.0")]
 pub const unsafe fn unreachable_unchecked() -> ! {
+    debug_assert!(false, "entered unreachable code");
```
Comment (camsteffen) [inline library/core/src/hint.rs:49]: just a suggestion, to be more informative
```suggestion
    debug_assert!(false, "called `unreachable_unchecked`");
```

**Comment by JohnCSimon** [comment]
ping from triage:
@ibraheemdev  
Returning to you to address build failures and after that, please switch back to `S-waiting-on-review`.
thanks.

@rustbot label: +S-waiting-on-author -S-waiting-on-review

**Comment by JohnCSimon** [comment]
Ping from triage: 
@ibraheemdev  can you post your status on this PR? It has sat idle for a while.

**Comment by ibraheemdev** [comment]
cc @RalfJung I'm not sure what to do about [this test](https://github.com/rust-lang/rust/blob/master/src/test/ui/consts/const_unsafe_unreachable_ub.rs), which is now panicking.

**Comment by RalfJung** [comment]
I guess the test should check for the panic then?

The problem will rather be that behavior differs depending on how the standard library is compiled. For codegen tests we have `// ignore-debug`, I wonder if that also works for ui tests?

**Comment by JohnCSimon** [comment]
@ibraheemdev 
Ping from triage: I'm closing this due to inactivity, Please reopen when you are ready to continue with this. Thanks for your contribution.

@rustbot label: +S-inactive

**Pull request closed without merging**